### DISCUSSION TITLE
Excluding WebHttpBindig from WSDL

### DISCUSSION THREAD

**Discussion by qcz** [CoreWCF/CoreWCF#1487] [Q&A] [answered]
I'm trying to migrate our WCF service to CoreWCF. Almost everything works perfectly. One problem I face is that the WebHttpBinding shows up in the WSDL.

I create the service as following:

```cs
builder.AddService<MyService>(serviceOptions =>
{
	serviceOptions.BaseAddresses.Add(new Uri("https://myhostname/MyService.svc"));
	serviceOptions.DebugBehavior.HttpHelpPageEnabled = true;
	serviceOptions.DebugBehavior.IncludeExceptionDetailInFaults = false;
});

builder.AddServiceEndpoint<MyService, IMyService>(
	new BasicHttpBinding()
	{
		Security = new() { Mode = BasicHttpSecurityMode.Transport },
	},
	"/soap",
	(endpointConfigurator) =>
	{
		// Skipped unneccessary extra configuration
	}
);

builder.AddServiceWebEndpoint<MyService, IMyService>(
	new WebHttpBinding()
	{
		Security = new() { Mode = WebHttpSecurityMode.Transport },
	},
	"/api",
	(endpointConfigurator) =>
	{
		// Skipped unneccessary extra configuration
	}
);

builder.ConfigureServiceHostBase<MyService>(host =>
{
	var smb = host.Description.Behaviors.Find<ServiceMetadataBehavior>();
	if (smb == null)
		host.Description.Behaviors.Add(new ServiceMetadataBehavior());

	smb.HttpsGetEnabled = true;
	smb.HttpsGetUrl = new System.Uri("/MyService.svc?wsdl", System.UriKind.Relative);
});
```

The outputted WSDL contains both the BasicHttpBinding and the WebHttpBinding, but the WebHttpBinding is not a valid soap service (eg. it shows up, but can't call it from SoapUI). It is not included in the WSDL generated by the original WCF app.
```xml
<wsdl:service name="MyService">
    <wsdl:port name="BasicHttpBinding_IMyService" binding="tns:BasicHttpBinding_IMyService">
        <soap:address location="https://myhostname/MyService.svc/soap"/>
    </wsdl:port>
    <wsdl:port name="WebHttpBinding_IMyService" binding="tns:WebHttpBinding_IMyService">
        <soap12:address location="https://myhostname/MyService.svc/api"/>
    </wsdl:port>
</wsdl:service>
```

**Comment by mconnew** [comment]
I thought WCF did generate a WSDL for WebHttpBinding, but when searching everywhere says it doesn't. I can't find the mechanism which suppresses it in WCF, but there is an easy workaround. Change your ConfigureServiceHostBase method to this and it should disable it being added to the WSDL:
```c#
builder.ConfigureServiceHostBase<MyService>(host =>
{
    var smb = host.Description.Behaviors.Find<ServiceMetadataBehavior>();
    if (smb == null)
        host.Description.Behaviors.Add(new ServiceMetadataBehavior());

    smb.HttpsGetEnabled = true;
    smb.HttpsGetUrl = new System.Uri("/MyService.svc?wsdl", System.UriKind.Relative);
    foreach (ServiceEndpoint endpoint in host.Description.Endpoints)
    {
        if (endpoint.Binding.GetType() == typeof(WebHttpBinding))
        {
            ServiceMetadataContractBehavior contractBehavior = ((KeyedByTypeCollection<IContractBehavior>)endpoint.Contract.ContractBehaviors).Find<ServiceMetadataContractBehavior>();
            if (contractBehavior != null)
            {
                contractBehavior.MetadataGenerationDisabled = true;
            }
            else
            {
                endpoint.Contract.ContractBehaviors.Add(new ServiceMetadataContractBehavior(true));
            }
        }
    }
});
```
This adds a special behavior to the endpoint which tells the ServiceMetadataBehavior to skip adding that endpoint to the wsdl.

**Reply by qcz** [reply]
Thanks for the tip! If I add this code (without any other modifications), the wsdl and singleWsdl endpoints cease to function:

- The service name is removed from the main service page (the "You have created a service." page). It says "Service" instead of "MyService Service" in the title.
- `?wsdl` and `?singleWsdl` pages return the main service page instead of the WSDL themselves.

**Reply by mconnew** [reply, marked as answer]
I missed that you were using the same contract interface for both bindings. This is perfectly fine and there's no problem with doing that, but it's the reason my suggestion didn't work. Because it adds the behavior to ServiceEndpoint.Contract, and both endpoints use the same contract, it gets added to both. You can either make a duplicate of your contract just for use with WebHttpBinding and have your service implement that too (and as the method names and signatures are identical, no code changes needed in your service other than declaring implementing the interface), and then use the approach I mentioned above, or you can wait for me to work out the code to yank it from the exported wsdl, which might be a day or two.

**Reply by qcz** [reply]
Thank you, the fixed version will be fine.

**Reply by mconnew** [reply]
I worked out how it's suppressed on .NET Framework. WebMessageEncodingBindingElement overrides an internal virtual method called `IsWsdlExportable`. The base class MessageEncodingBindingElement returns true, but the overridden method returns false. This method isn't public in WCF. We didn't implement `IWsdlExportExtension.ExportEndpoint` in WebMessageEncodingBindingElement in CoreWCF, but it is implemented in WCF. I had a look at the implementation and ExportEndpoint should be implemented still as it removes some things from the WSDL that shouldn't be there as the EnvelopeVersion is None. It looks like even the WCF implementation could have done with some more iteration as WebHttpBinding will pollute your WSDL with any data types that are only in the WebHttpBinding endpoint. On the whole you'll actually get a cleaner wsdl having separate contracts and using the ServiceMetadataContractBehavior mechanism than you get on WCF. Although in your case, as the methods are identical between them, it's a non-issue for your specific case.
